src/typed-geometry/functions/objects/intersection.hh

```diff
@@ -2269,4 +2327,617 @@ template <class ScalarT>
     return true;
 }
 
+template <class ScalarT>
+[[nodiscard]] constexpr optional<segment<3, ScalarT>> intersection(plane<3, ScalarT> const& plane, triangle<3, ScalarT> const& t)
+{
+    // classify vertices
+    auto sign_v1 = signed_distance(t.pos0, plane) < 0 ? false : true;
+    auto sign_v2 = signed_distance(t.pos1, plane) < 0 ? false : true;
+    auto sign_v3 = signed_distance(t.pos2, plane) < 0 ? false : true;
+
+    // exclude some degenerate cases? e.g. vertices of triangle on same positions, angle constraints..
+
+    if (sign_v1 == sign_v2 && sign_v2 == sign_v3) // no intersection (early out)
+        return {};
+
+    // isolated vertex
+    bool iv = (sign_v1 == sign_v2) ? sign_v3 : (sign_v1 == sign_v3) ? sign_v2 : sign_v1;
+
+    pos<3, ScalarT> i1, i2;
+
+    // intersection exists (exactly 2 vertices on one side of the plane and exactly 1 vertex on the other side)
+    if (iv == sign_v1)
+    {
+        i1 = intersection(segment<3, ScalarT>(t.pos0, t.pos1), plane).value();
+        i2 = intersection(segment<3, ScalarT>(t.pos0, t.pos2), plane).value();
+    }
+    else if (iv == sign_v2)
+    {
+        i1 = intersection(segment<3, ScalarT>(t.pos0, t.pos1), plane).value();
+        i2 = intersection(segment<3, ScalarT>(t.pos1, t.pos2), plane).value();
+    }
+    else if (iv == sign_v3)
+    {
+        i1 = intersection(segment<3, ScalarT>(t.pos0, t.pos2), plane).value();
+        i2 = intersection(segment<3, ScalarT>(t.pos1, t.pos2), plane).value();
+    }
+    else
+        return {};
+
+    return tg::segment3(i1, i2);
+}
+
+template <class ScalarT>
+[[nodiscard]] constexpr optional<segment<3, ScalarT>> intersection(triangle<3, ScalarT> const& t, plane<3, ScalarT> const& plane)
+{
+    return intersection(plane, t);
+}
+
+template <class ScalarT>
+[[nodiscard]] constexpr bool intersects(plane<3, ScalarT> const& plane, triangle<3, ScalarT> const& t)
+{
+    tg::array<pos<3, ScalarT>, 3> triangle_pos = {t.pos0, t.pos1, t.pos2};
+    float sign = 0;
```

Comment:
ScalarT probably

Reply:
yep. Do I have type-cast _0_ in this case to ScalarT when comparing _sign_ :
**sign == ScalarT(0)** ?

Reply:
we usually do that yes. the reason is that ScalarT can be quite sophisticated and does not have to be in {int, float, double}. it can be tg::complex. it can be some bigint, bigrat. it can be a tracing type to collect execution info. not all these types can be implicitly constructed (or tested) against integer literals. but we expect an explicit construction. so if ScalarT is more sophisticated, it can happen that `x == 0` does not compile, while `x == ScalarT(0)`  does.

Reply:
Note: it's not hard enforced because those types are a bit niche and we often only use a tg API subset with them